commands/scan.md

```diff
@@ -9,19 +9,203 @@ Since these commands allow for incremental iteration, returning only a small num
+
+## Calling SCAN with a corrupted cursor
+
+Calling `SCAN` with a broken, negative, out of range, or otherwise invalid cursor, will result in undefined behavior but never in a crash. What will be undefined is that the guarantees about the returned elements can no longer be ensured by the `SCAN` implementation.
+
+The only valid cursors to use are:
+
+* The cursor value of 0 when starting an iteration.
+* The cursor returned by the previous call to SCAN in order to continue the iteration.
+
+## Guarantee of termination
+
+The `SCAN` algorithm is guaranteed to terminate only if the size of the iterated collection remains bounded to a given maximum size, otherwise iterating a collection that always grows may result into `SCAN` to never terminate a full iteration.
+
+This is easy to see intuitively: if the collection grows there is more and more work to do in order to visit all the possible elements, and the ability to terminate the iteration depends on the number of calls to `SCAN` and its COUNT option value compared with the rate at which the collection grows.
+
+## Why SCAN may return all the items of an aggregate data type in a single call?
+
+In the `COUNT` option documentation, we state that sometimes this family of commands may return all the elements of a Set, Hash or Sorted Set at once in a single call, regardless of the `COUNT` option value. The reason why this happens is that the cursor-based iterator can be implemented, and is useful, only when the aggregate data type that we are scanning is represented as a hash table. However Redis uses a [memory optimization](/topics/memory-optimization) where small aggregate data types, until they reach a given amount of items or a given max size of single elements, are represented using a compact single-allocation packed encoding. When this is the case, `SCAN` has no meaningful cursor to return, and must iterate the whole data structure at once, so the only sane behavior it has is to return everything in a call.
+
+However once the data structures are bigger and are promoted to use real hash tables, the `SCAN` family of commands will resort to the normal behavior. Note that since this special behavior of returning all the elements is true only for small aggregates, it has no effects on the command complexity or latency. However the exact limits to get converted into real hash tables are [user configurable](/topics/memory-optimization), so the maximum number of elements you can see returned in a single call depends on how big an aggregate data type could be and still use the packed representation.
+
+Also note that this behavior is specific of `SSCAN`, `HSCAN` and `ZSCAN`. `SCAN` itself never shows this behavior because the key space is always represented by hash tables.
+
+## Further reading
+
+For more information on `SCAN` please refer to the [The Redis Keyspace](/docs/manual/keyspace) tutorial.
```

Comment:
So if we delete most of the same text from the keyspace tutorial (which was move there from scan command docs here: https://github.com/redis/redis-doc/commit/882c184514f88f048df3d773e300a39882773e9d#diff-6875ae7c6d2de818737cec8e3e3c0221294ab337dab94fbe7c05ad66e1f789ce) there is not much more information about scan there. Maybe instead something like this:

```suggestion
For more information about managing keys, please refer to the [The Redis Keyspace](/docs/manual/keyspace) tutorial.
```

And to avoid having duplicate information on two places, we can delete Scan basic usage, scan guarantees, etc. from this tutorial and instead link to the SCAN command from there.